src/sql_generator.cpp

```diff
@@ -0,0 +1,342 @@
+                   const std::string &schema_name) {
+  std::ostringstream out;
+  out << "SELECT schema_name FROM information_schema.schemata WHERE "
+         "catalog_name="
+      << KeywordHelper::WriteQuoted(db_name, '\'')
+      << " AND schema_name=" << KeywordHelper::WriteQuoted(schema_name, '\'');
+
+  auto query = out.str();
+  mdlog::info("schema_exists: " + query);
+  auto result = con.Query(query);
+
+  if (result->HasError()) {
+    throw std::runtime_error("Could not find whether schema <" + schema_name +
+                             "> exists in database <" + db_name +
+                             ">: " + result->GetError());
+  }
+  return result->RowCount() > 0;
+}
+
+bool table_exists(duckdb::Connection &con, const table_def &table) {
+  std::ostringstream out;
+  out << "SELECT table_name FROM information_schema.tables WHERE table_catalog="
+      << KeywordHelper::WriteQuoted(table.db_name, '\'') << " AND table_schema="
+      << KeywordHelper::WriteQuoted(table.schema_name, '\'')
+      << " AND table_name="
+      << KeywordHelper::WriteQuoted(table.table_name, '\'');
+
+  auto query = out.str();
+  mdlog::info("table_exists: " + query);
+  auto result = con.Query(query);
+
+  if (result->HasError()) {
+    throw std::runtime_error("Could not find whether table <" +
+                             compute_absolute_table_name(table) +
+                             "> exists: " + result->GetError());
+  }
+
+  return result->RowCount() > 0;
+}
+
+void create_schema(duckdb::Connection &con, const std::string &db_name,
+                   const std::string &schema_name) {
+  auto query = "CREATE schema " + KeywordHelper::WriteQuoted(schema_name, '\'');
+  mdlog::info("create_schema: " + query);
+  con.Query(query);
+}
+
+void create_table(duckdb::Connection &con, const table_def &table,
+                  const std::vector<column_def> &columns) {
+  const std::string absolute_table_name = compute_absolute_table_name(table);
+  std::ostringstream ddl;
+  ddl << "CREATE OR REPLACE TABLE " << absolute_table_name << " (";
+
+  for (const auto &col : columns) {
+    ddl << KeywordHelper::WriteQuoted(col.name, '"') << " "
+        << duckdb::EnumUtil::ToChars(col.type);
+    if (col.primary_key) {
+      ddl << " PRIMARY KEY";
+    }
+    ddl << ", "; // DuckDB allows trailing commas
+  }
+
+  ddl << ")";
+
+  auto query = ddl.str();
+  mdlog::info("create_table: " + query);
+
+  auto result = con.Query(query);
+  if (result->HasError()) {
+    throw std::runtime_error("Could not create table <" + absolute_table_name +
+                             ">" + result->GetError());
+  }
+}
+
+std::vector<column_def> describe_table(duckdb::Connection &con,
+                                       const table_def &table) {
+  // TBD is_identity is never set, used is_nullable=no temporarily but really
+  // should use duckdb_constraints table.
+  std::ostringstream sql;
+  sql << "SELECT column_name, data_type, is_nullable == 'NO' FROM "
```

Comment:
duckdb is forgiving of `==` but technically sql is `=`
```suggestion
  sql << "SELECT column_name, data_type, is_nullable = 'NO' FROM "
```